### PR TITLE
KNNMetricCallback implemented and tested.

### PR DESCRIPTION
**Pull request by smivv**
## Description

As discussed [here](https://github.com/catalyst-team/catalyst/issues/459), added KNNMetricCallback to measure KNN accuracy/precision/recall/f1-score during the evaluation. Also I can add some comments and docs if needed.

Code was fully tested, here the link to [colab](https://colab.research.google.com/drive/12Fad5kts6_9PxquBhFZl6y4MdPeL_ZFw).

## Related Issue

[Issue](https://github.com/catalyst-team/catalyst/issues/459)

## Type of Change

- [ ] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [x] I have read the [Code of Conduct](https://github.com/catalyst-team/catalyst/blob/master/CODE_OF_CONDUCT.md) document.
- [x] I have read the [Contributing](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md) guide.
- [x] I have checked the code-style using `make check-style`.
- [x] I have written the docstring in Google format for all the methods and classes that I used.
- [x] I have checked the docs using `make check-docs`.

### CODE REVIEW

File: catalyst/contrib/callbacks/knn.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+            for i, value in enumerate(values):
+                postfix = self.class_names[i] \
+                    if self.class_names is not None else str(i)
+                loader_values[f"{self.prefix}/{postfix}"] = value
+
+        self.reset_cache()
+
+    def on_epoch_end(self, state: RunnerState):
+
+        for s in ["valid", "test"]:
```
Comment (Scitator) [inline catalyst/contrib/callbacks/knn.py]: here should be some more general solution :)

File: catalyst/contrib/callbacks/knn.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+
+        super().__init__(CallbackOrder.Metric)
+        self.prefix = prefix
+        self.features_key = input_key
+        self.targets_key = output_key
+
+        self.num_classes = num_classes
+        self.class_names = class_names
+
+        if metric_fn == 'accuracy':
```
Comment (Scitator) [inline catalyst/contrib/callbacks/knn.py]: quotes :)

File: catalyst/contrib/callbacks/knn.py (hunk 1 above)
Comment (smivv) [reply catalyst/contrib/callbacks/knn.py]: Didn't thought about that, added additional parameter :)

File: catalyst/contrib/callbacks/knn.py (hunk 2 above)
Comment (smivv) [reply catalyst/contrib/callbacks/knn.py]: Fixed :)

File: catalyst/contrib/callbacks/knn.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+    """
+    A callback that returns single metric on `state.on_batch_end`
+    """
+
+    def __init__(
+            self,
+            input_key: str = "logits",
+            output_key: str = "targets",
+            prefix: str = "knn",
+            num_classes: int = 2,
```
Comment (Scitator) [inline catalyst/contrib/callbacks/knn.py:25]: I suggest to add assert for `> 1` check

File: catalyst/contrib/callbacks/knn.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+                    if self.class_names is not None else str(i)
+                loader_values[f"{self.prefix}/{postfix}"] = value
+
+        self.reset_cache()
+
+    def on_epoch_end(self, state: RunnerState):
+
+        for s in self.cv_loader_names:
+            if s in self.sets:
+                y_true, y_pred = self._knn(self.sets["train"], self.sets[s])
```
Comment (Scitator) [inline catalyst/contrib/callbacks/knn.py]: there are exists cases when we don't have "train" loader-name (it can be named like "train_1" for example), so this should be parametrized

File: catalyst/contrib/callbacks/knn.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+            # and it is not possible to put everything in memory, so we split
+            except MemoryError:
+                print(f"Memory error with {self.n_folds} fold, trying more.")
+                self.n_folds *= 2
+                result = None
+
+        return result
+
+
+__all__ = ["KNNMetricCallback"]
```
Comment (Scitator) [inline catalyst/contrib/callbacks/knn.py:252]: also, please follow our Class methods organisation, like:
```
class A:
  def inner1(): ...
  def inner2(): ...
  def inner3():
    based on inner1/inner2
```

so, please place all method you use in some "main" function before this "main" function

**Comment by Scitator** [comment]
also looks, like there are some codestyle issue yet
please follow our [contribution best practices](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md#contribution-best-practices) to make codestyle checks easier :)

File: catalyst/contrib/callbacks/knn.py (hunk 3 above)
Comment (smivv) [reply catalyst/contrib/callbacks/knn.py:25]: Added :)

File: catalyst/contrib/callbacks/knn.py (hunk 4 above)
Comment (smivv) [reply catalyst/contrib/callbacks/knn.py]: Added this functionality in `cv_loader_names` parameter.

File: catalyst/contrib/callbacks/knn.py (hunk 5 above)
Comment (smivv) [reply catalyst/contrib/callbacks/knn.py:252]: Done :)

**Comment by smivv** [comment]
@Scitator, code style also should be ok now :)
